src/Analyzer.Cli/CommandLineParser.cs

```diff
@@ -0,0 +1,139 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Invocation;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Templates.Analyzer.Cli
+{
+    internal class CommandLineParser
+    {
+        RootCommand rootCommand;
+
+        /// <summary>
+        /// Constructor for the command line parser. Sets up the command line API. 
+        /// </summary>
+        public CommandLineParser()
+        {
+            SetupCommandLineAPI();
+        }
+
+        /// <summary>
+        /// Invoke the command line API using the provided arguments. 
+        /// </summary>
+        /// <param name="args">Arguments sent in via the command line</param>
+        /// <returns>A Task that executes the command handler</returns>
+        public async Task InvokeCommandLineAPIAsync(string[] args)
+        {
+            await rootCommand.InvokeAsync(args).ConfigureAwait(false);
+        }
+
+        private RootCommand SetupCommandLineAPI()
+        {
+            // Command line API is setup using https://github.com/dotnet/command-line-api
+            // Create a root command 
+            rootCommand = new RootCommand();
+            rootCommand.Description = "Analyze Azure Resource Manager (ARM) Templates for security and best practice issues.";
+
+            // It has two commands - analyze-template and analyze-directory
+            rootCommand.AddCommand(SetupAnalyzeTemplateCommand());
+
+            rootCommand.AddCommand(SetupAnalyzeDirectoryCommand());
+            
+            return rootCommand;
+        }
+
+        private Command SetupAnalyzeTemplateCommand()
+        {
+            // Setup analyze-template 
+            Command analyzeTemplateCommand = new Command("analyze-template");
+
+            Option<FileInfo> templateOption = new Option<FileInfo>(
+                    "--template-file-path",
+                    "The ARM template to analyze")
+            {
+                IsRequired = true
+            };
+            templateOption.AddAlias("-t");
+
+            analyzeTemplateCommand.AddOption(templateOption);
+
+            Option<FileInfo> parameterOption = new Option<FileInfo>(
+                 "--parameters-file-path",
+                 "The parameter file to use when parsing the specified ARM template");
+            parameterOption.AddAlias("-p");
+
+            analyzeTemplateCommand.AddOption(parameterOption);
+
+            analyzeTemplateCommand.AddOption(SetupOutputFileOption());
+
+            analyzeTemplateCommand.Handler = CommandHandler.Create<FileInfo, FileInfo>((templateFilePath, parametersFilePath) =>
+            {
+                try
+                {
+                    Core.TemplateAnalyzer templateAnalyzer = new Core.TemplateAnalyzer(File.ReadAllText(templateFilePath.FullName), parametersFilePath == null ? null : File.ReadAllText(parametersFilePath.FullName));
```

Comment:
For testing purposes, instead of directly creating a `TemplateAnalyzer` here, pass in a builder to the constructor of this class that can create the `TemplateAnalyzer` with a template and parameters.  For now, we can keep the builder simple.  Something like this should be fine:
``` C#
// Constructor
public CommandLineParser(Func<string, string, TemplateAnalyzer> analyzerBuilder) { ... }

// Creating CommandLineParser
new CommandLineParser((template, parameters) => new TemplateAnalyzer(template, parameters));
```
And replace here with:
``` suggestion
... = analyzerBuilder(File.ReadAllText(...), ...);
```

Then in the tests, pass in a builder that will create a mock.

Reply:
Tagging issue #28.